components/gfx/display_list/mod.rs

```diff
@@ -347,97 +217,44 @@ impl DisplayList {
         paint_context.draw_target.set_transform(&old_transform);
     }
 
-    pub fn find_stacking_context<'a>(&'a self,
-                                     stacking_context_id: StackingContextId)
-                                     -> Option<&'a StackingContext> {
-        fn find_stacking_context_in_stacking_context<'a>(stacking_context: &'a StackingContext,
-                                                         stacking_context_id: StackingContextId)
-                                                          -> Option<&'a StackingContext> {
-            if stacking_context.id == stacking_context_id {
-                return Some(stacking_context);
-            }
-
-            for kid in stacking_context.children() {
-                let result = find_stacking_context_in_stacking_context(kid, stacking_context_id);
-                if result.is_some() {
-                    return result;
+    fn draw_with_state<'a>(&'a self,
+                           traversal: &mut DisplayListTraversal,
+                           paint_context: &mut PaintContext,
+                           transform: &Matrix4D<f32>,
+                           subpixel_offset: &Point2D<Au>,
+                           tile_rect: Option<Rect<Au>>) {
+        loop {
```

Comment:
How about `while let Some(item) = traversal.next() {`? Then just make the `PopStackingContext` case break (or return). It'd save a level of indentation here.